### PR TITLE
Process

### PR DESCRIPTION
**Pull request by imkiva**
DO NOT code review now, as some features are still missing:

* [x] wait_for() returns exit code
* [x] interrupt() force terminates the process
* [x] is_exited() returns whether the process is running

and are untested on win32 platforms:

* [x] mpp_impl::terminate_process()
* [x] environment variables
* [x] mpp_impl::wait_for()

I just opened this PR to keep track of my work progress.

### CODE REVIEW

File: mozart++/system/process.hpp [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+
+        int wait_for() {
+            if (_exit_code >= 0) {
+                return _exit_code;
+            }
+            _exit_code = mpp_impl::wait_for(_info);
+            return _exit_code;
+        }
+
+        bool is_exited() const {
```
Comment (mikecovlee) [inline mozart++/system/process.hpp]: I think this is not very reasonable. If user use this function but don't use wait_for, the exit code will never change.

File: mozart++/system/process.hpp [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+            }
+            return *this;
+        }
+
+        template <typename Container>
+        process_builder &arguments(const Container &c) {
+            if (_startup._cmdline.size() <= 1) {
+                std::copy(c.begin(), c.end(), std::back_inserter(_startup._cmdline));
+            } else {
+                // invalid operation, do nothing
```
Comment (mikecovlee) [inline mozart++/system/process.hpp]: Consider throwing an exception?

File: mozart++/system/process.hpp [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+        for (const auto &s : startup._cmdline) {
+            ss << s << " ";
+        }
+
+        std::string command = ss.str();
+
+        if (!CreateProcess(nullptr, const_cast<char *>(command.c_str()),
+            nullptr, nullptr, true, CREATE_NO_WINDOW, nullptr,
+            startup._cwd.c_str(), &si, &pi)) {
+            mpp::throw_ex<mpp::runtime_error>("unable to fork subprocess");
```
Comment (mikecovlee) [inline mozart++/system/process.hpp]: Lack of CloseHandle here.

File: mozart++/system/process.hpp [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+            info._stdin = pstdin[PIPE_WRITE];
+            info._stdout = pstdout[PIPE_READ];
+            info._stderr = pstderr[PIPE_READ];
+
+            // on *nix systems, fork() doesn't create threads to run process
+            info._tid = FD_INVALID;
+        }
+    }
+
+#endif
```
Comment (mikecovlee) [inline mozart++/system/process.hpp]: In fact, considering the possibilities of throwing an other exception inside this function(such as constructor of std::string), I strongly recommend using RAII to protect pipes here.

File: mozart++/system/file.hpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include <cstring>
+
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <unistd.h>
+#endif
+
+namespace mpp {
+#ifdef _WIN32
```
Comment (mikecovlee) [inline mozart++/system/file.hpp]: In Mozart++, you can use `MOZART_PLATFORM_WIN32` instead.

File: mozart++/system/process.hpp (hunk 4 above)
Comment (imkiva) [reply mozart++/system/process.hpp]: It's not easy to apply RAII to these pipe fds. 
First of all, sometimes fds[0] and fds[1] should be both closed (when errored), but sometimes one of fds[0] and fds[1] should remain open util process exits.

The best way to ensure no resource leak is applying try-catch in its caller and close pipes.

File: mozart++/system/file.hpp (hunk 5 above)
Comment (imkiva) [reply mozart++/system/file.hpp]: ok

File: mozart++/system/process.hpp (hunk 4 above)
Comment (mikecovlee) [reply mozart++/system/process.hpp]: ok. that is better.

File: mozart++/system/process.hpp (hunk 1 above)
Comment (imkiva) [reply mozart++/system/process.hpp]: See new approach

File: mozart++/core/optional.hpp [hunk 6]
```diff
@@ -184,7 +184,7 @@ namespace mpp {
          * @param o default value
          * @return reference to the object or to the default object
          */
-        T &get_or(T &o) {
+        T &get_or(T &&o) {
```
[the remark this answers is not in the document]
Comment (imkiva) [reply mozart++/core/optional.hpp]: See new approach

**Pull request merged**